accounts/abi/unpack.go

```diff
@@ -113,17 +112,25 @@ func forEachUnpack(t Type, output []byte, start, size int) (interface{}, error)
 		return nil, fmt.Errorf("abi: invalid type in array/slice unpacking stage")
 	}
 
-	for i, j := start, 0; j*32 < len(slice); i, j = i+32, j+1 {
-		// this corrects the arrangement so that we get all the underlying array values
-		if t.Elem.T == ArrayTy && j != 0 {
-			i = start + t.Elem.Size*32*j
-		}
+	for i, j := start, 0; j < size; j++ {
 		inter, err := toGoType(i, *t.Elem, output)
 		if err != nil {
 			return nil, err
 		}
+
+		reflectedInter := reflect.ValueOf(inter)
+
+		//Although we just did the reverse, pack it, to get the length of the actual element.
+		//Getting the length directly from the "toGoType" would be way better,
+		// but it requires some refactoring to get it return the *consumed* length.
+		interPacked, err  := t.Elem.pack(reflectedInter)
+		if err != nil {
+			return nil, err
+		}
+		i += len(interPacked)
```

Comment:
By packing it, and using the length as the increment, it guarantees correct element alignment (given that packing works correctly). However, getting a "consumed" size from the `toGoType` call would be *much* more efficient. I decided to go with this inefficient solution first, since changing `toGoType` affects other code.